### ISSUE TITLE
Implementation Plan for MCP Server

### ISSUE THREAD

**Issue by yoheimuta** [yoheimuta/protolint#468]
## Overview

This document outlines the implementation plan for adding Model Context Protocol (MCP) server functionality to protolint. The implementation will focus on local file system usage, utilizing stdio transport for communication between MCP clients and the protolint server.

## 1. Basic Design

1. **Add MCP Reporter**:
   - Create a new reporter in `linter/report/reporters/mcpReporter.go`
   - Extend the existing reporter mechanism to include MCP-compatible formatting

2. **Add --mcp Flag**:
   - Add a `--mcp` option to the protolint command
   - Enable MCP server mode when this flag is present

3. **Implement MCP Server**:
   - Create an `internal/mcp` directory for server-related code
   - Implement request/response protocol handling
   - Implement communication via stdin/stdout

## 2. Detailed Implementation

<details>
<summary>Code</summary>

### 2.1 MCP Reporter Implementation

Add a new reporter in `mcpReporter.go` that formats failures into MCP-compatible JSON:

```go
package reporters

import (
    "encoding/json"
    "fmt"
    "io"

    "github.com/yoheimuta/protolint/linter/report"
)

// MCPReporter prints failures in MCP friendly JSON format.
type MCPReporter struct{}

// Report writes failures to w in MCP friendly format.
func (r MCPReporter) Report(w io.Writer, fs []report.Failure) error {
    // Group failures by file
    fileFailures := make(map[string][]map[string]interface{})
    
    for _, failure := range fs {
        filePath := failure.Pos().Filename
        
        failureInfo := map[string]interface{}{
            "rule_id":  failure.RuleID(),
            "message":  failure.Message(),
            "line":     failure.Pos().Line,
            "column":   failure.Pos().Column,
            "severity": failure.Severity(),
        }
        
        fileFailures[filePath] = append(fileFailures[filePath], failureInfo)
    }
    
    // Convert to array of results
    var fileResults []map[string]interface{}
    for filePath, failures := range fileFailures {
        fileResults = append(fileResults, map[string]interface{}{
            "file_path": filePath,
            "failures":  failures,
        })
    }
    
    result := map[string]interface{}{
        "results": fileResults,
    }
    
    bs, err := json.MarshalIndent(result, "", "  ")
    if err != nil {
        return err
    }
    
    _, err = fmt.Fprintln(w, string(bs))
    return err
}
```

Next, add the `mcp` reporter to the `GetReporter` function in `internal/cmd/subcmds/lint/reporterFlag.go`:

```go
// GetReporter returns a reporter from the specified key.
func GetReporter(value string) (report.Reporter, error) {
    rs := map[string]report.Reporter{
        // Existing reporters
        ...
        "mcp": reporters.MCPReporter{},  // Add MCP reporter
        ...
    }
    // ...
}
```

#### protocol.go

```go
package mcp

import (
    "encoding/json"
)

// Request represents an MCP request
type Request struct {
    Type    string          `json:"type"`
    ID      string          `json:"id"`
    Payload json.RawMessage `json:"payload"`
}

// Response represents an MCP response
type Response struct {
    Type    string      `json:"type"`
    ID      string      `json:"id"`
    Payload interface{} `json:"payload"`
}

// ListToolsResponse represents the response for list_tools request
type ListToolsResponse struct {
    Tools []ToolInfo `json:"tools"`
}

// ToolInfo represents information about a tool
type ToolInfo struct {
    Name        string      `json:"name"`
    Description string      `json:"description"`
    Schema      interface{} `json:"schema,omitempty"`
}

// CallToolPayload represents the payload for call_tool request
type CallToolPayload struct {
    Name      string          `json:"name"`
    Arguments json.RawMessage `json:"arguments"`
}

// CallToolResponse represents the response for call_tool request
type CallToolResponse struct {
    Result interface{} `json:"result"`
}
```

#### server.go

```go
package mcp

import (
    "encoding/json"
    "fmt"
    "io"
    "os"
    
    "github.com/yoheimuta/protolint/internal/osutil"
)

// Server represents an MCP server
type Server struct {
    tools []Tool
}

// NewServer creates a new MCP server
func NewServer() *Server {
    return &Server{
        tools: []Tool{
            NewLintFilesTool(),
            // Other tools can be added in the future
        },
    }
}

// Run starts the MCP server
func (s *Server) Run(stdout, stderr io.Writer) osutil.ExitCode {
    fmt.Fprintf(stderr, "protolint MCP server is running. cwd: %s\n", getCurrentDir())
    
    decoder := json.NewDecoder(os.Stdin)
    encoder := json.NewEncoder(stdout)
    
    for {
        var request Request
        if err := decoder.Decode(&request); err != nil {
            if err == io.EOF {
                // Normal termination
                return osutil.ExitSuccess
            }
            fmt.Fprintf(stderr, "Error decoding request: %v\n", err)
            return osutil.ExitInternalFailure
        }
        
        response := s.handleRequest(&request)
        if err := encoder.Encode(response); err != nil {
            fmt.Fprintf(stderr, "Error encoding response: %v\n", err)
            return osutil.ExitInternalFailure
        }
    }
}

// handleRequest handles a single MCP request
func (s *Server) handleRequest(req *Request) *Response {
    switch req.Type {
    case "list_tools":
        return s.handleListTools(req)
    case "call_tool":
        return s.handleCallTool(req)
    default:
        return &Response{
            Type: "error",
            ID:   req.ID,
            Payload: map[string]string{
                "message": fmt.Sprintf("Unknown request type: %s", req.Type),
            },
        }
    }
}

// handleListTools handles list_tools request
func (s *Server) handleListTools(req *Request) *Response {
    toolInfos := make([]ToolInfo, 0, len(s.tools))
    for _, tool := range s.tools {
        toolInfos = append(toolInfos, tool.GetInfo())
    }
    
    return &Response{
        Type: "list_tools_response",
        ID:   req.ID,
        Payload: &ListToolsResponse{
            Tools: toolInfos,
        },
    }
}

// handleCallTool handles call_tool request
func (s *Server) handleCallTool(req *Request) *Response {
    var payload CallToolPayload
    if err := json.Unmarshal(req.Payload, &payload); err != nil {
        return &Response{
            Type: "error",
            ID:   req.ID,
            Payload: map[string]string{
                "message": fmt.Sprintf("Invalid payload: %v", err),
            },
        }
    }
    
    // Find the tool
    var tool Tool
    for _, t := range s.tools {
        if t.GetInfo().Name == payload.Name {
            tool = t
            break
        }
    }
    
    if tool == nil {
        return &Response{
            Type: "error",
            ID:   req.ID,
            Payload: map[string]string{
                "message": fmt.Sprintf("Tool not found: %s", payload.Name),
            },
        }
    }
    
    // Execute the tool
    result, err := tool.Execute(payload.Arguments)
    if err != nil {
        return &Response{
            Type: "error",
            ID:   req.ID,
            Payload: map[string]string{
                "message": fmt.Sprintf("Tool execution failed: %v", err),
            },
        }
    }
    
    return &Response{
        Type: "call_tool_response",
        ID:   req.ID,
        Payload: &CallToolResponse{
            Result: result,
        },
    }
}

// getCurrentDir returns the current working directory
func getCurrentDir() string {
    dir, err := os.Getwd()
    if err != nil {
        return "<unknown>"
    }
    return dir
}
```

#### tools.go

```go
package mcp

import (
    "bytes"
    "encoding/json"
    "fmt"
    
    "github.com/yoheimuta/protolint/lib"
)

// Tool defines the interface for MCP tools
type Tool interface {
    GetInfo() ToolInfo
    Execute(args json.RawMessage) (interface{}, error)
}

// LintFilesTool is a tool for linting Proto files
type LintFilesTool struct{}

// NewLintFilesTool creates a new LintFilesTool
func NewLintFilesTool() *LintFilesTool {
    return &LintFilesTool{}
}

// GetInfo returns the tool information
func (t *LintFilesTool) GetInfo() ToolInfo {
    return ToolInfo{
        Name:        "lint-files",
        Description: "Lint Protocol Buffer files using protolint",
        Schema: map[string]interface{}{
            "type": "object",
            "properties": map[string]interface{}{
                "files": map[string]interface{}{
                    "type": "array",
                    "items": map[string]interface{}{
                        "type": "string",
                    },
                    "description": "List of file paths to lint",
                },
                "config_path": map[string]interface{}{
                    "type": "string",
                    "description": "Path to protolint config file",
                },
                "fix": map[string]interface{}{
                    "type": "boolean",
                    "description": "Fix lint errors if possible",
                },
            },
            "required": []string{"files"},
        },
    }
}

// LintFilesArgs represents arguments for lint-files tool
type LintFilesArgs struct {
    Files      []string `json:"files"`
    ConfigPath string   `json:"config_path,omitempty"`
    Fix        bool     `json:"fix,omitempty"`
}

// Execute runs the lint-files tool
func (t *LintFilesTool) Execute(args json.RawMessage) (interface{}, error) {
    var lintArgs LintFilesArgs
    if err := json.Unmarshal(args, &lintArgs); err != nil {
        return nil, fmt.Errorf("invalid arguments: %v", err)
    }
    
    if len(lintArgs.Files) == 0 {
        return nil, fmt.Errorf("no files specified")
    }
    
    // Construct command line arguments
    cmdArgs := []string{}
    
    // Use MCP reporter
    cmdArgs = append(cmdArgs, "--reporter", "mcp")
    
    if lintArgs.ConfigPath != "" {
        cmdArgs = append(cmdArgs, "--config_path", lintArgs.ConfigPath)
    }
    
    if lintArgs.Fix {
        cmdArgs = append(cmdArgs, "--fix")
    }
    
    // Add files at the end
    cmdArgs = append(cmdArgs, lintArgs.Files...)
    
    // Capture output to parse it
    var outputBuffer bytes.Buffer
    var errorBuffer bytes.Buffer
    
    // Run lint command
    err := lib.Lint(cmdArgs, &outputBuffer, &errorBuffer)
    
    // Determine exit code based on error
    exitCode := 0
    if err != nil {
        if err == lib.ErrLintFailure {
            exitCode = 1
        } else {
            exitCode = 2
            // Return error information if internal error occurred
            return map[string]interface{}{
                "exit_code": exitCode,
                "error":     err.Error(),
                "stderr":    errorBuffer.String(),
            }, nil
        }
    }
    
    // Parse the JSON output from MCP reporter
    var result map[string]interface{}
    if err := json.Unmarshal(outputBuffer.Bytes(), &result); err != nil {
        return nil, fmt.Errorf("failed to parse lint output: %v", err)
    }
    
    // Add exit code to result
    result["exit_code"] = exitCode
    
    return result, nil
}
```

### 2.3 Adding MCP Flag

Update `internal/cmd/cmd.go` to handle the `--mcp` flag:

```go
const (
    // Existing flags
    ...
    // MCP flag
    mcpFlag = "--mcp"
)

// Do function extension
func Do(
    args []string,
    stdout io.Writer,
    stderr io.Writer,
) osutil.ExitCode {
    // Check for --mcp flag
    for _, arg := range args {
        if arg == mcpFlag {
            // Start MCP server mode
            server := mcp.NewServer()
            return server.Run(stdout, stderr)
        }
        // Existing flag checks
        ...
    }
    
    // Existing processing
    ...
}
```
</details>

## 3. Operational Flow

1. User runs `protolint --mcp` command
2. The `cmd.Do` function detects the `--mcp` flag and starts the MCP server
3. MCP server reads requests from stdin and writes responses to stdout
4. MCP client (e.g., Claude Desktop) sends requests to the server and receives responses

### Request Example:

```json
{
  "type": "list_tools",
  "id": "request-1234"
}
```

### Response Example:

```json
{
  "type": "list_tools_response",
  "id": "request-1234",
  "payload": {
    "tools": [
      {
        "name": "lint-files",
        "description": "Lint Protocol Buffer files using protolint",
        "schema": {
          "type": "object",
          "properties": {
            "files": {
              "type": "array",
              "items": { "type": "string" },
              "description": "List of file paths to lint"
            },
            "config_path": {
              "type": "string",
              "description": "Path to protolint config file"
            },
            "fix": {
              "type": "boolean",
              "description": "Fix lint errors if possible"
            }
          },
          "required": ["files"]
        }
      }
    ]
  }
}
```

## 4. Integration with Claude Desktop

To use with Claude Desktop, edit the configuration file:

```json
{
  "mcpServers": {
    "protolint": {
      "command": "protolint",
      "args": ["--mcp"],
      "type": "stdio"
    }
  }
}
```

## 5. Advantages

1. **Reuse of Existing Code**:
   - Leverages existing reporter mechanism for output formatting
   - Uses lib.Lint to utilize existing lint functionality

2. **Clear Responsibility Separation**:
   - MCP reporter handles MCP format JSON output
   - MCP server handles request/response processing
   - Transport layer (stdio) handles communication

3. **Future Extensibility**:
   - Additional tools can be added
   - Reporter improvements will reflect in MCP

By adopting this approach, we can effectively add MCP server functionality with minimal changes. Additionally, by maximizing the reuse of existing code, maintenance becomes easier.
